File: src/main/java/org/pcha/benthos/language/BloblangParserDefinition.java
package org.pcha.benthos.language;

import com.intellij.lang.ASTNode;
import com.intellij.lang.ParserDefinition;
import com.intellij.lang.PsiParser;
import com.intellij.lexer.Lexer;
import com.intellij.openapi.project.Project;
import com.intellij.psi.FileViewProvider;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IFileElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;
import org.pcha.benthos.language.parser.BloblangParser;
import org.pcha.benthos.language.psi.BloblangFile;
import org.pcha.benthos.language.psi.BloblangTypes;

public class BloblangParserDefinition implements ParserDefinition {

    public static final TokenSet WHITE_SPACES = TokenSet.create(TokenType.WHITE_SPACE);
    public static final TokenSet COMMENTS = TokenSet.create(BloblangTypes.COMMENT);

    public static final IFileElementType FILE = new IFileElementType(BloblangLanguage.INSTANCE);

    @Override
    public @NotNull Lexer createLexer(Project project) {
        return new BloblangLexerAdapter();
    }

    @Override
    public @NotNull TokenSet getWhitespaceTokens() {
        return WHITE_SPACES;
    }

    @Override
    public @NotNull TokenSet getCommentTokens() {
        return COMMENTS;
    }

    @Override
    public @NotNull TokenSet getStringLiteralElements() {
        return TokenSet.EMPTY;
    }

    @Override
    public @NotNull PsiParser createParser(Project project) {
        return new BloblangParser();
    }

    @Override
    public @NotNull IFileElementType getFileNodeType() {
        return FILE;
    }

    @Override
    public @NotNull PsiFile createFile(@NotNull FileViewProvider viewProvider) {
        return new BloblangFile(viewProvider);
    }

    @Override
    public @NotNull PsiElement createElement(ASTNode node) {
        return BloblangTypes.Factory.createElement(node);
    }
}

File: src/main/kotlin/com/github/pcha/intellij/bloblang/BloblangParserDefinition.kt
package com.github.pcha.intellij.bloblang

import com.github.pcha.intellij.bloblang.parser.BloblangParser
import com.github.pcha.intellij.bloblang.psi.BloblangFile
import com.github.pcha.intellij.bloblang.psi.BloblangTokenSets
import com.github.pcha.intellij.bloblang.psi.BloblangTypes
import com.intellij.lang.ASTNode
import com.intellij.lang.ParserDefinition
import com.intellij.lang.PsiParser
import com.intellij.lexer.Lexer
import com.intellij.openapi.project.Project
import com.intellij.psi.FileViewProvider
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.tree.IFileElementType
import com.intellij.psi.tree.TokenSet
import org.jetbrains.annotations.NotNull

internal class BloblangParserDefinition : ParserDefinition {
    @NotNull
    override fun createLexer(project: Project?): Lexer {
        return BloblangLexerAdapter()
    }

    @NotNull
    override fun getCommentTokens(): TokenSet {
        return BloblangTokenSets.COMMENTS
    }

    @NotNull
    override fun getStringLiteralElements(): TokenSet {
        return BloblangTokenSets.STRING_LITERALS
    }

    @NotNull
    override fun createParser(project: Project?): PsiParser {
        return BloblangParser()
    }

    @NotNull
    override fun getFileNodeType(): IFileElementType {
        return FILE
    }

    @NotNull
    override fun createFile(@NotNull viewProvider: FileViewProvider): PsiFile {
        return BloblangFile(viewProvider)
    }

    @NotNull
    override fun createElement(node: ASTNode?): PsiElement {
        return BloblangTypes.Factory.createElement(node)
    }

    companion object {
        val FILE: IFileElementType = IFileElementType(BloblangLanguage.INSTANCE)
    }
}